File: lib/lintus/report.rb
# frozen_string_literal: true

module Lintus
  # Everything a run produced: what was checked, what was flagged, what was skipped, what broke.
  class Report
    Checked = Struct.new(:path, :answers, keyword_init: true)
    Skipped = Struct.new(:path, :reason, keyword_init: true)
    Failure = Struct.new(:path, :error, keyword_init: true)

    attr_reader :checked, :offenses, :skipped, :failures

    def initialize
      @checked = []
      @offenses = []
      @skipped = []
      @failures = []
      @mutex = Mutex.new
    end

    # `answers` maps every rule asked about the file to the probability it got,
    # offense or not, so a run can be read for confidence and not just verdicts.
    def record_checked(path, offenses, answers = {})
      synchronize do
        checked << Checked.new(path: path, answers: answers)
        self.offenses.concat(offenses)
      end
    end

    def record_skipped(path, reason)
      synchronize { skipped << Skipped.new(path: path, reason: reason) }
    end

    def record_failure(path, error)
      synchronize { failures << Failure.new(path: path, error: error) }
    end

    def sorted_offenses = offenses.sort_by { |offense| [offense.path, offense.rule.id] }

    # 2 when a file could not be checked, 1 when there are offenses, else 0.
    def exit_status
      return 2 if failures.any?

      offenses.any? ? 1 : 0
    end

    private

    def synchronize(&) = @mutex.synchronize(&)
  end
end

File: src/report.rs
//! Everything a run produced: what was checked, what was flagged, what was skipped, what broke.

/// A file the model answered about, with the probability each rule got,
/// offense or not, so a run can be read for confidence and not just verdicts.
#[derive(Debug, Clone, PartialEq)]
pub struct Checked {
    pub path: String,
    pub answers: Vec<(String, f64)>,
}

/// A rule the model flagged on a file, with the probability it gave.
#[derive(Debug, Clone, PartialEq)]
pub struct Offense {
    pub path: String,
    pub rule: String,
    pub message: String,
    pub noul: f64,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Skipped {
    pub path: String,
    pub reason: String,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Failure {
    pub path: String,
    pub error: String,
}

#[derive(Debug, Default, Clone, PartialEq)]
pub struct Report {
    pub checked: Vec<Checked>,
    pub offenses: Vec<Offense>,
    pub skipped: Vec<Skipped>,
    pub failures: Vec<Failure>,
}

impl Report {
    /// Puts every list in a stable order, whatever order the files finished in:
    /// by path, and offenses by path then rule.
    pub fn sort(&mut self) {
        self.checked.sort_by(|a, b| a.path.cmp(&b.path));
        self.offenses.sort_by(|a, b| (&a.path, &a.rule).cmp(&(&b.path, &b.rule)));
        self.skipped.sort_by(|a, b| a.path.cmp(&b.path));
        self.failures.sort_by(|a, b| a.path.cmp(&b.path));
    }

    /// 2 when a file could not be checked, 1 when there are offenses, else 0.
    pub fn exit_status(&self) -> i32 {
        if !self.failures.is_empty() {
            2
        } else if !self.offenses.is_empty() {
            1
        } else {
            0
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn offense(path: &str, rule: &str) -> Offense {
        Offense { path: path.into(), rule: rule.into(), message: String::new(), noul: 0.9 }
    }

    #[test]
    fn exit_status() {
        let mut report = Report::default();
        assert_eq!(report.exit_status(), 0);

        report.offenses.push(offense("a.rb", "r"));
        assert_eq!(report.exit_status(), 1);

        report.failures.push(Failure { path: "b.rb".into(), error: "boom".into() });
        assert_eq!(report.exit_status(), 2);
    }

    #[test]
    fn offenses_sort_by_path_then_rule() {
        let mut report = Report {
            offenses: vec![offense("b.rb", "no_sleep"), offense("a.rb", "no_sleep"), offense("a.rb", "documented")],
            ..Report::default()
        };
        report.sort();

        let order: Vec<_> = report.offenses.iter().map(|o| (o.path.as_str(), o.rule.as_str())).collect();
        assert_eq!(order, [("a.rb", "documented"), ("a.rb", "no_sleep"), ("b.rb", "no_sleep")]);
    }
}
